### PR TITLE
Improve ItemStack#editMeta

### PR DESCRIPTION
**Pull request by Machine-Maker**
I think some people might find this handy, I think someone mentioned something similar in discord today, and I thought I'd make it happen. Should preserve existing behavior of ItemStack#editMeta because all meta is an instance of ItemMeta.

### CODE REVIEW

**Review by MiniDigger — APPROVED** [review]
I like

File: patches/api/0309-ItemStack-editMeta.patch [hunk 1]
```diff
@@ -27,9 +27,28 @@ index b4d55e0d5e605cb6980b555acd42d16e1e65c196..f3fa8f7f021464aa9d463c14550487cc
...
++     * (such as {@link #getItemMeta()}, {@link #addItemFlags(ItemFlag...)}, {@link #lore()}, etc.)
++     * from inside the consumer is disallowed and will produce undefined results or exceptions.
++     * </p>
++     *
++     * @param metaClass the type of meta to edit
++     * @param consumer the meta consumer
++     * @param <M> the meta type
++     * @return {@code true} if the edit was successful, {@code false} otherwise
++     */
++    public <M extends ItemMeta> boolean editMeta(Class<M> metaClass, final @NotNull java.util.function.Consumer<? super M> consumer) {
```
Comment (MiniDigger) [inline patches/api/0309-ItemStack-editMeta.patch]: class is missing @NotNull, which breaks the test. make it final while at it.

Comment (Machine-Maker) [reply patches/api/0309-ItemStack-editMeta.patch]: Done

File: patches/api/0308-ItemStack-editMeta.patch [hunk 2]
```diff
@@ -27,9 +27,28 @@ index 86bd9f14de5c1ff3d797955be1af56e5efcac884..59a026d80b0a0a4890becf98efdbe532
...
++     * @param metaClass the type of meta to edit
++     * @param consumer the meta consumer
++     * @param <M> the meta type
++     * @return {@code true} if the edit was successful, {@code false} otherwise
++     */
++    public <M extends ItemMeta> boolean editMeta(final @NotNull Class<M> metaClass, final @NotNull java.util.function.Consumer<? super M> consumer) {
 +        final ItemMeta meta = this.getItemMeta();
-+        if (meta != null) {
-+            consumer.accept(meta);
++        if (meta != null && metaClass.isInstance(meta)) {
```
Comment (lynxplay) [inline patches/api/0308-ItemStack-editMeta.patch]: Could remove the `meta != null` check, as this just doubles logic at this point.

**Review by kashike — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Comment by Owen1212055** [comment]
I'm curious, is there any reason why the "normal" editMeta method isn't something like this?
```java
public <T extends ItemMeta> boolean editMeta(final @NotNull java.util.function.Consumer<T> consumer) {
		return editMeta(ItemMeta.class, consumer);
}
```

This would allow us to specify the type in the lambda
```java
item.editMeta((SkullMeta meta) -> {});
```

Trying to do this with the current method causes a syntax error "Cannot infer functional interface type".

**Pull request merged**